src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    }
     return nbytes;
 }
 
-struct fsobj *find(const char *path, struct fsobj *root)
+void file_rewinddir()
 {
-    assert(root);
-    assert(strlen(root->name) < strlen(path));
-    while (1) {
-        assert(strcmp(root->name, path) !=0 );
-        assert(strspn(path, root->name) >= strlen(root->name));
-        if (!root->child) return NULL;
+    if (!openfile) return;
+    if (openfile->parent == NULL) return;
+    assert(openfile->parent->child);
+    openfile = openfile->parent->child;
+    return;
+}
 
-        root = root->child;
+struct fsobj *file_find(const char *path)
+{
+    //paths should always begin with '/' dir names do not
+    if (path[0] != '/') return NULL;
+    path++;
+
+    assert(path);
+    if(!openfs || !openfs->root) return NULL;
+    struct fsobj *root = openfs->root;
+    assert(root && root->name);
+    if (strlen(root->name) > strlen(path)) return NULL;
+
+    while (1) {
 
+        // iterate through linked list of children
         while (strspn(path, root->name) < strlen(root->name)) {
             if (!strcmp(root->name, path)) return root;
             if (!root->next) return NULL;
             root = root->next;
         }
 
+        //root must now be parent of final location
+        if (!strcmp(root->name, path)) return root;
+        assert(strspn(path, root->name) >= strlen(root->name));
+        if (!root->child) return NULL;
+
+        root = root->child;
+        while (path[0] != '/') {
+            path++;
+        }
+        path++;
+
     }
     PANIC("Unreachable Statement");
     return NULL;
 }
 
+int32 file_mkdir(const char *path, uint32 access)
+{
+    assert(path);
+    struct fsobj *o;
+    if (newfile_get(&o)) return OS_FS_ERR_DRIVE_NOT_CREATED;
+    o->name = path_to_name((char *) path);
+    o->type = FSOBJ_DIR;
+    o->size = 0;
+    if (file_insert(o, (char *) path)) return OS_FS_ERR_PATH_INVALID;
+
+    return OS_FS_SUCCESS;
+}
+
+int32 file_rmdir(const char *path)
+{
+    assert(path);
+    struct fsobj *root = file_find(path);
+    if (root == NULL) return OS_FS_ERROR;
+    struct fsobj *cur;
+
+    while (root->child) {
+
+        // if cur is the last leaf in a list
+        if (cur->next == NULL && cur->child == NULL) {
+            if (cur->prev != NULL) {
+                assert(cur->prev->next == cur);
+                cur = cur->prev;
+                rm(cur->next);
+            }
+            else {
+                assert(cur->parent->child == cur);
+                cur = cur->parent;
+                rm(cur->child);
+            }
+        } else if (cur->child != NULL) {
+            cur = cur->child;
+        } else { //cur->next !=NULL
+            cur = cur->next;
+        }
+    }
+
+    rm(root);
+    return OS_FS_SUCCESS;
+}
+
+int32 file_creat(const char *path)
+{
+    assert(path);
+    struct fsobj *o;
+    if (newfile_get(&o)) return OS_FS_ERR_DRIVE_NOT_CREATED;
+    o->name = (char *) path;
```

Comment:
If you have to cast away const, don't take const parameters. This needs to be fixed throughout.